jupyterhub/files/hub/jupyterhub_config.py

```diff
@@ -74,11 +73,26 @@ def camelCaseify(s):
         cfg_key = camelCaseify(trait)
     set_config_if_not_none(c.JupyterHub, trait, 'hub.' + cfg_key)
 
-c.JupyterHub.ip = os.environ['PROXY_PUBLIC_SERVICE_HOST']
-c.JupyterHub.port = int(os.environ['PROXY_PUBLIC_SERVICE_PORT'])
+# Note that *_SERVICE_PORT and *_SERVICE_HOST environment variables are set by
+# Kubernetes to help processes in containers discover Kubernetes services.
+#
+# ref: https://kubernetes.io/docs/concepts/services-networking/service/#environment-variables
+
+# hub_bind_url configures what the JupyterHub process within the hub pod's
+# container should listen to.
+hub_container_port = 8081
+# Tornados httpserver can only listens to either IPv4 or IPv6 traffic, which
+# deviates from Linux's standard behavior. Due to this, JupyterHub currently
+# does not support dual-stacks setup where either IPv4 or IPv6 can be used.
+ipv6_enabled_cluster = ':' in os.environ['KUBERNETES_SERVICE_HOST']
+if ipv6_enabled_cluster
+    c.JupyterHub.hub_bind_url = f'http://[::]:{hub_container_port}'
+else:
+    c.JupyterHub.hub_bind_url = f'http://0.0.0.0:{hub_container_port}'
```

Comment:
@minrk @stv0g I was considering this limitation about only being able to listen either on IPv4 or IPv6? I could not find any clear confirmation on this.

Perhaps though JupyterHub ends up limiting us even though Tornado's httpserver does. What JupyterHub ends up doing to actively listen is this call [here](https://github.com/jupyterhub/jupyterhub/blob/6ecda96dd636dfdd0cf34fcf6c1fc44942779c01/jupyterhub/app.py#L2645), but before this call, some processing of the configuration is made.

I wonder if it would work to set this to f`http://:{hub_container_port}` instead, as my understanding is that tornado would support both IPv4 or IPv6 if a blank string or None value was used for the IP.

Hmmm... But looking more in depth, i see that there seem to be three ways to start Tornado, and we seem to follow the "single process" pattern. See: https://www.tornadoweb.org/en/stable/httpserver.html#http-server.

I find that the `listen` function we use is documented [here, in the TCPServer class which HTTPServer is derived from](https://www.tornadoweb.org/en/stable/tcpserver.html#tornado.tcpserver.TCPServer.listen).

The interesting part is defined in the [`bind` function documentation](https://www.tornadoweb.org/en/stable/tcpserver.html#tornado.tcpserver.TCPServer.bind).

> Address may be an empty string or None to listen on all available interfaces. Family may be set to either socket.AF_INET or socket.AF_INET6 to restrict to IPv4 or IPv6 addresses, otherwise both will be used if available.

It sounds to me, that as long as we pass an empty string when calling `listen`, we should be able to listen to both IPv4 and IPv6 - could this be correct?

Reply:
I tried to use `http://:8081` and that worked well for IPv4, but based on the Tornado documentation I think it will also work for IPv6 as well.